limit number of comments displayed, option to delete comments

hjohns28 opened the pull request:
parent pull request: #22

hjohns28 committed:
limit number of comments displayed, option to delete comments

kristen-law commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,14 +63,23 @@ function previousExperience() {
   selectedExperience--;
 }
 
+var number = 0;
```
Try to be more specific when naming your variables. It isn't clear what `number` is referring to here. You could rename it to something like `numComments` or even `numCommentsDisplayed` so it's clear what the value means.

kristen-law commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,14 +63,23 @@ function previousExperience() {
 
+var number = 0;
+
+function storeNumberComments() {
```
Similarly, I think this function name is a little misleading. From what I can tell, this isn't storing the number of comments, but instead the storing the number of comments to display to the user and then populating the comment section. I'd recommend renaming this too.

kristen-law commented on portfolio/src/main/webapp/index.html:
```diff
@@ -30,6 +30,19 @@ <h1>Heidi Johnson's Portfolio</h1>
+          
+          <label for="number-comments">Number of comments:</label>
+            <select onchange="storeNumberComments()" name="number-comments" id="number-comments">
+              <option value="0">0</option>
```
You don't really need to specify the `value="x"` part here. If the value attribute isn't specified in the `<option>` tag, the content is passed as the value instead. Since the values are the same as the content here, you should still get the same result.

kristen-law commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,14 +63,23 @@ function previousExperience() {
   selectedExperience--;
 }
 
+var number = 0;
```
I also don't think this variable needs to have a global scope. Think about how you might be able to accomplish this, perhaps by adding a parameter to the `getServerComments` function.

hjohns28 replied:
This is a good point, thank you, I'm just having some trouble with one aspect. I know if I initialize number inside storeNumberComments(), I could pass it to getServerComments(number) as a parameter. However in my index.html, I have <body onload="getServerComments()" so that when the user posts a new comment, the page re-fetches the comments so that the new one shows up. When this happens, though, the value that was selected in the drop down (for the number of comments to display) is lost because it resets. Since I would now need to go through storeNumberComments() to get to getServerComments() for the onload=, I need to figure out a way to preserve the selected value in the drop down, but I'm having trouble doing this. That is why I had given the variable global scope, but I will continue researching how to do this. If you have any suggestions on how to preserve the value, that'd be great. In the meantime I'll keep working at it!

hjohns28 replied:
Sorry for that long message, I just wanted to add that I've tried to solve it by using local storage so that when the page reloads, the drop down gets the value that was selected from storage, which is what almost all answers say to do online, but I keep getting errors.

hjohns28 committed:
give number variable more meaningful name

hjohns28 committed:
change to more meaningful names, change scope of 'number'

hjohns28 committed:
merge last commit with fixes

hjohns28 committed:
fix merge conflict I didn't address in last commit

kristen-law commented on portfolio/src/main/webapp/index.html:
```diff
@@ -32,15 +32,24 @@ <h1>Heidi Johnson's Portfolio</h1>
-              <option value="3">3</option>
-              <option value="4">4</option>
-              <option value="5">5</option>
+            <select onclick="getNumCommentsSelected()" name="number-comments" id="number-comments">
```
Using `onclick` will call the function when you first click on the select element instead of when you choose the option. You probably want to keep it as `onchange`.

hjohns28 replied:
Would it be okay to leave it as onclick? I changed it because when I had it as onchange, it wasn't refreshing when you selected a new value, I think because of the <script> part that stores the value onchange. It still works like before, and fetches the correct number of comments when you select a new number. Thanks for any advice!

hjohns28 replied:
Nevermind to my above comment, I got it fixed with onchange, thank you!

kristen-law commented on portfolio/src/main/webapp/index.html:
```diff
@@ -32,15 +32,24 @@ <h1>Heidi Johnson's Portfolio</h1>
-          
+          <script>
+            document.getElementById('number-comments').onchange = function() {
+              localStorage.setItem('selectedtem', document.getElementById('number-comments').value);
```
Is this supposed to be `selectedItem`?

hjohns28 replied:
yes, whoops. Thank you.

hjohns28 committed:
fix typo in name of selectedItem

FabriceCastel commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -34,23 +34,41 @@
+
+  private int getNumberComments(HttpServletRequest request) {
+    String numberCommentsString = request.getParameter("number-comments");
+    int numberComments = 0;
```
Might be a bit nicer to default to a non-zero number of requested comments, but that's more of a matter of personal preference/design

FabriceCastel commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -34,23 +34,41 @@
+
+  }
+
+  private int getNumberComments(HttpServletRequest request) {
```
`getRequestedCommentCount` or something similar for the name to be more meaningful/easier to understand

FabriceCastel commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -34,23 +34,41 @@
 
-        String comment = text;
-        commentSection.add(comment);
+    int numberComments = getNumberComments(request);
```
Small nit: `numberComments` -> `commentCount` or `numberOfComments`

FabriceCastel commented on portfolio/src/main/webapp/index.html:
```diff
@@ -30,6 +30,28 @@ <h1>Heidi Johnson's Portfolio</h1>
+              <option>5</option>
+            </select> 
+          <script>
+            document.getElementById('number-comments').onchange = function() {
+              localStorage.setItem('selectedItem', document.getElementById('number-comments').value);
+            };
+
+            if (localStorage.getItem('selectedItem')) {
+              document.getElementById('number-comments').options[localStorage.getItem('selectedItem')].selected = true;
+            }
```
Is there a reason for this script to be embedded directly in your HTML page as opposed to living in a separate .js file?

FabriceCastel commented on portfolio/src/main/webapp/index.html:
```diff
@@ -30,6 +30,28 @@ <h1>Heidi Johnson's Portfolio</h1>
+            </select> 
+          <script>
+            document.getElementById('number-comments').onchange = function() {
+              localStorage.setItem('selectedItem', document.getElementById('number-comments').value);
```
`localStorage` seems potentially excessive here, is there a reason this was used as opposed to a simple variable or `sessionStorage`?

FabriceCastel commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,14 +63,21 @@ function previousExperience() {
 
-function getServerComments() {
-  fetch('/data').then(response => response.json()).then((commentSection) => {
+function getNumCommentsSelected() {
+  var comments = document.getElementById('number-comments');
+  var numCommentsSelected = comments.options[comments.selectedIndex].value;
+  getServerComments(numCommentsSelected);
+}
```
This ties into Kristen's comment from earlier but you could save yourself a bit of work here by having this method accept an index parameter and then in your HTML, doing this:

```
<select onchange="getNumCommentsSelected(this.selectedIndex)" (....)>
```

hjohns28 replied:
Thanks for all the feedback, this all helped a lot!

hjohns28 committed:
revise names, change to sessionStorage, change onclick to onchange

hjohns28 committed:
Merge branch 'commentoptionsbranch' of https://github.com/hjohns28/step into commentoptionsbranch

FabriceCastel commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,21 +63,51 @@ function previousExperience() {
+  fetch('/delete-data', {method: 'POST'}).then(getServerComments(0));
+}
+
+function storeSelectedValue(selected) {
+  sessionStorage.setItem('selectedValue', selected.value);
+  if (sessionStorage.getItem('selectedValue')) {
+      document.getElementById('number-comments').options[sessionStorage.getItem('selectedValue')].selected = true;
+  }
+}
```
The only difference between `sessionStorage` and using a variable here would be that `sessionStorage` persists when you change pages on your site and then come back to this page, whereas the variable would get reset. Was this a deliberate design choice on your part or did you end up using (`localStorage` earlier, now `sessionStorage`) that for some other reason, eg. the previously embedded JS code in the HTML wouldn't work without it?

FabriceCastel replied:
This is not so much a request for changes, it's more of an FYI and to ask if you knew this/intended for this beforehand.

hjohns28 replied:
I used storage because when the page reloaded (after a new comment was added for example), I wanted the selected drop down to stay. Not so much for switching pages on the site, but more so for refreshing this particular page. Are there drawbacks to this / do you think this might be unnecessary? (you don't have to answer, but I'd be curious and happy to know your input)

FabriceCastel replied:
If you can avoid reloading the page, that would be best. If it seems like more than a brief effort, then you can skip it for the time being. In general, it's good to avoid full page reloads where possible. Modern websites tend to fetch data from the backend piece by piece and iteratively update the UI in small chunks. This means less data is transferred between the backend and the frontend and has the user spending less time waiting for things to load/update.

hjohns28 replied:
Okay, that makes sense and is good to know, thank you. I might wait on making this change so that I can make sure I deploy last week's work and can start this week's tutorial, but I will remember this.

FabriceCastel commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,21 +63,51 @@ function previousExperience() {
+  }
+}
+
+function getStoredValue() {
+  var selectedValue = sessionStorage.getItem('selectedValue');
+  if (selectedValue === null) {
+    selectedValue = 1;
+  }
+  getServerComments(selectedValue);
 }
```
This function does more than its name seems to suggest

FabriceCastel commented on portfolio/src/main/webapp/script.js:
```diff
@@ -63,21 +63,51 @@ function previousExperience() {
+    individualComments.appendChild(createComment(element));
   });
+  if (sessionStorage.getItem('selectedValue') === null){
+    document.getElementById('number-comments').options[1].selected = true;
```
Is this necessary? It looks like this option will always be selected when you first load the page, since you've added this to the HTML

hjohns28 replied:
Since the sessionStorage only comes into play onchange, when you first load the page the default value hadn't been stored yet (at least that's what seemed to be happening, and adding this fixed it). Would it be okay to leave it in?

FabriceCastel replied:
As far as I can tell, what is happening here is that you're manually setting the `<option>` in the UI to be selected, which then triggers an `onchange` call and so on, with the goal of ultimately setting the `selectedValue` in `sessionStorage`.

What I'm saying is that your default `<option>` is already marked as `selected` in the HTML, so you could skip that step altogether and replace that line with `sessionStorage.setItem(....)`, unless I'm missing something here.

hjohns28 replied:
I see what you're saying, my apologies, that makes sense.

FabriceCastel commented on portfolio/src/main/webapp/index.html:
```diff
@@ -30,6 +30,19 @@ <h1>Heidi Johnson's Portfolio</h1>
           <ul id="comments-container"></ul>
+          
+          <label for="number-comments">Number of comments:</label>
+            <select onchange="getNumCommentsSelected(this); storeSelectedValue(this)" name="number-comments" id="number-comments">
```
You should always wrap functionality like this in a single method call instead of writing out multiple expressions in an `onchange` etc. field, so here you should implement a single method to call rather than call both `getNumCommentsSelected` and `storeSelectedValue` separately.

FabriceCastel commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -34,23 +34,41 @@
-        commentSection.add(comment);
+    int commentCount = getRequestedCommentCount(request);
+    
+    ArrayList<String> limitedComments = new ArrayList<String>();
+    for (int i = 0; i < commentCount; i++) {
+        limitedComments.add(commentSection.get(i));
     }
```
You don't need to do this work here, you can instead add the number of results you want directly in the query you have above using SELECT TOP, LIMIT or equivalent

hjohns28 committed:
start to address fixes, not done yet

hjohns28 committed:
combine getNumCommentsSelected and storeSelectedValue, reduce fetching comments to one loop

start to address fixes, not done yet

combine getNumCommentsSelected and storeSelectedValue, reduce fetching comments to one loop

hjohns28 committed:
combine getNumCommentsSelected and storeSelectedValue, reduce fetching comments to one loop

hjohns28 committed:
change to store default selected value if no value has been selected

hjohns28 committed:
Merge branch 'commentoptionsbranch' of https://github.com/hjohns28/step into commentoptionsbranch

FabriceCastel reviewed (approved):
LGTM

kristen-law commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -34,23 +34,44 @@
+        commentSection.add(text);
     }
     
+    //ArrayList<String> limitedComments = new ArrayList<String>();
+    //for (int i = 0; i < commentCount; i++) {
+      //  limitedComments.add(commentSection.get(i));
+    //}
```
Make sure you delete this code if you're not using it.

hjohns28 replied:
yes, sorry. I just deleted it on a different branch, I'll delete it here too.

hjohns28 committed:
delete unused code that was commented out
